File: Session_04_GreedyAlgorithms/multi_language_implementations/activity_selection.js
const readline = require('readline');

function selectActivities(activities) {
    const sortedActivities = [...activities].sort((a, b) => a.finish - b.finish);
    console.log("Sorted Activities by Finish Time:");
    for (let act of sortedActivities) {
        console.log(`  Activity ID ${act.id}: [${act.start}, ${act.finish}]`);
    }

    console.log("\nScheduling Decisions:");
    let selectedCount = 0;
    let lastFinishTime = sortedActivities[0].finish;
    console.log(`  Selected Activity ID ${sortedActivities[0].id}: [${sortedActivities[0].start}, ${sortedActivities[0].finish}] (First to finish)`);
    selectedCount++;

    for (let i = 1; i < sortedActivities.length; i++) {
        const act = sortedActivities[i];
        if (act.start >= lastFinishTime) {
            console.log(`  Selected Activity ID ${act.id}: [${act.start}, ${act.finish}] (Starts after last finish ${lastFinishTime})`);
            lastFinishTime = act.finish;
            selectedCount++;
        } else {
            console.log(`  Skipped Activity ID ${act.id}: [${act.start}, ${act.finish}] (Conflict)`);
        }
    }
    console.log(`\nTotal selected compatible activities: ${selectedCount}`);
}

function main() {
    const activities = [
        {id: 1, start: 5, finish: 9},
        {id: 2, start: 1, finish: 2},
        {id: 3, start: 3, finish: 4},
        {id: 4, start: 0, finish: 6},
        {id: 5, start: 5, finish: 7},
        {id: 6, start: 8, finish: 9}
    ];
    selectActivities(activities);
    console.log();

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    rl.question("Press Enter to exit...", () => {
        rl.close();
    });
}

main();

File: Session_04_GreedyAlgorithms/multi_language_implementations/activity_selection.py
import sys

class Activity:
    def __init__(self, activity_id, start, finish):
        self.id = activity_id
        self.start = start
        self.finish = finish

def select_activities(activities):
    # Sort activities by finish time
    activities.sort(key=lambda x: x.finish)
    
    print("Sorted Activities by Finish Time:")
    for act in activities:
        print(f"  Activity ID {act.id}: [{act.start}, {act.finish}]")
    print("\nScheduling Decisions:")
    
    selected_count = 0
    last_finish_time = activities[0].finish
    print(f"  Selected Activity ID {activities[0].id}: [{activities[0].start}, {activities[0].finish}] (First to finish)")
    selected_count += 1
    
    for i in range(1, len(activities)):
        act = activities[i]
        if act.start >= last_finish_time:
            print(f"  Selected Activity ID {act.id}: [{act.start}, {act.finish}] (Starts after last finish {last_finish_time})")
            last_finish_time = act.finish
            selected_count += 1
        else:
            print(f"  Skipped Activity ID {act.id}: [{act.start}, {act.finish}] (Conflict)")
            
    print(f"\nTotal selected compatible activities: {selected_count}")

def main():
    activities = [
        Activity(1, 5, 9),
        Activity(2, 1, 2),
        Activity(3, 3, 4),
        Activity(4, 0, 6),
        Activity(5, 5, 7),
        Activity(6, 8, 9)
    ]
    select_activities(activities)
    print()
    input("Press Enter to exit...")

if __name__ == "__main__":
    main()
